Prevent queries without at least one scalar aggregator & allow disabling this restriction

philkjacobs opened the pull request:
* updated the reloading config `jsonData` to add another field to it -- `enforceScalarSetting`. Documented this in the config UI.
* Added an allowlist for scalar aggregator names -- I don't love this solution, but didn't have anything cleaner that doesn't involve cascading changes to every single `Aggregator` class. Included the `grep` command used to generate this.
* Added `overrideScalar` field to `KairosDBTarget`, updated codepaths that create and copy these to ensure they would maintain the field and default to `false`
* Added checkbox 2-way bound to this `KairosDBTarget` field alongside copy explaining its usage.
* Updated datasource.ts to register and propagate the new config setting on datasource initialization.
* Updated `TargetValidator` to return a boolean + reason instead of just a boolean. This means we can now include customized error messages for failed validations. We can extend this to resolve similar problems/validations we want to error out for.
* Updated existing `TargetValidator` validations to have an error message.

<img width="1677" alt="enforceScalarAggOSSScreenshot" src="https://user-images.githubusercontent.com/2497883/85628306-c9dcf380-b624-11ea-9de4-870bc0b01c04.png">

vjkoskela commented on dist/beans/aggregators/aggregators.ts:
```diff
@@ -10,37 +10,54 @@ import {ScaleAggregator} from "./scale_aggregator";
     new RangeAggregator("sum"),
+];
+
+export const AGGREGATORS = RANGE_AGGREGATORS.concat([
```
If an aggregator is not in the `RANGE_AGGREGATORS` list does that mean it is not considered to return a scalar? Or perhaps you could elaborate on the what these two categorizations mean (via a comment)

philkjacobs replied:
They're unrelated sets. I didn't think entirely reworking the Aggregator datamodel is a change worth making here. 

We could register each of these properties as a boolean (scalar, range) for every Aggregator, but as-is, some of these aggregators are little more than a name string.

I think this change is one of the ones that has not yet been merged (and isn't part of my change).

philkjacobs replied:
Resolving as I've removed these unrelated changes.

vjkoskela commented on dist/beans/aggregators/aggregators.ts:
```diff
@@ -10,37 +10,54 @@ import {ScaleAggregator} from "./scale_aggregator";
-    "last", "least_squares", "max", "min", "gaps", "merge", "sum", "movingWindow"];
+// Loosely generated from:
+// tslint:disable-next-line
+// grep -H -q "doubleDataPointFactory" <path-to-kairosdb-extensions>/kairosdb-extensions/src/main/java/io/inscopemetrics/kairosdb/aggregators
```
This limits the list of aggregators to those supported by histograms. The Kairosdb-datasource is actually independent of histograms. You should probably also look at the aggregators in the main package as well:

https://github.com/kairosdb/kairosdb

vjkoskela commented:
First, to be clear, this change is _backwards compatible_ for user behavior because it is behind a feature flag which defaults to `false`.

Second, the goal of this change (when the feature is enabled) is to prevent user's from making queries without an aggregator. There are a few specialized cases where this is necessary, for example, when visualizing a heat map. However, at this time the intent cannot be inferred from the plugin.

Instead when enabled the feature requires a user to "sign off" on user an aggregator-less query. The Metrics Portal proxy will inject a merge aggregator when configured to do so. Switching both (the MP behavior and this KDB-DS behavior) to default is something that can be discussed separately.

@philkjacobs can you provide some screenshots?

Also, it's typically bad form to link to private resources in a public pull request.

philkjacobs commented:
> @philkjacobs can you provide some screenshots?
Added one.
> Also, it's typically bad form to link to private resources in a public pull request.
Fixed -- I hadn't noticed it when copying the diff summary over.
